Client/Frontend/Home/TopSitesPanel.swift

```diff
@@ -136,16 +138,18 @@ class TopSitesPanel: UIViewController {
     }
 
     private func deleteHistoryTileForSite(site: Site, atIndexPath indexPath: NSIndexPath) {
-        func reloadThumbnails() {
-            self.profile.history.getTopSitesWithLimit(self.layout.thumbnailCount)
-                .uponQueue(dispatch_get_main_queue()) { result in
+        func reloadThumbnails() -> Success {
+            return self.profile.history.getTopSitesWithLimit(self.layout.thumbnailCount)
+                .bindQueue(dispatch_get_main_queue()) { result in
                     self.deleteOrUpdateSites(result, indexPath: indexPath)
```

Comment:
It looks like the problem is caused by asynchronously updating the UI based on the index after doing the requery. The index is fragile since any requery during the async update can cause the indices to change.

Rather than doing the query and then updating the UI, what if we delete the item immediately from the collection (in `didRemoveThumbnail`), then fill in the gaps after the query is done? Theoretically, that means we wouldn't need to pass around `indexPath` anymore, which can go stale as we bounce between threads.